### ISSUE TITLE
SSL verify configuration behaviour and option

### ISSUE THREAD

**Issue by andydavidson** [bootc/nrpe-ng#2]
Hi,

I am struggling to debug why I am seeing this behaviour:

 - the nrpe-ng -H hostname matches the CN of the server certificate
 - I have turned off validation in the config file for my dev environment anyway.

Is this a CA validation issue?

Debian 9.2

```
root@inception:/home/andy# /usr/lib/nagios/plugins/check_nrpe_ng -H test -c check_load -C /etc/nagios/check_nrpe_ng.cfg --debug
[   ('args', None),
    ('command', 'check_load'),
    ('config_file', '/etc/nagios/check_nrpe_ng.cfg'),
    ('debug', True),
    ('host', 'test'),
    ('port', 59546),
    ('ssl_ca_file', ''),
    ('ssl_cert_file', '/etc/ssl/certs/ssl-cert-snakeoil.pem'),
    ('ssl_key_file', '/etc/ssl/private/ssl-cert-snakeoil.key'),
    ('ssl_verify_server', False),
    ('timeout', 10),
    ('timeout_unknown', False)]
Traceback (most recent call last):
  File "/usr/lib/nagios/plugins/check_nrpe_ng", line 11, in <module>
    load_entry_point('nrpe-ng==0.1.2', 'console_scripts', 'check_nrpe_ng')()
  File "/usr/lib/python3/dist-packages/nrpe_ng/client.py", line 197, in main
    return Client().run()
  File "/usr/lib/python3/dist-packages/nrpe_ng/client.py", line 177, in run
    conn.request(**req)
  File "/usr/lib/python3.5/http/client.py", line 1107, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.5/http/client.py", line 1152, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.5/http/client.py", line 1103, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/usr/lib/python3.5/http/client.py", line 877, in send
    self.connect()
  File "/usr/lib/python3.5/http/client.py", line 1261, in connect
    server_hostname=server_hostname)
  File "/usr/lib/python3.5/ssl.py", line 385, in wrap_socket
    _context=self)
  File "/usr/lib/python3.5/ssl.py", line 760, in __init__
    self.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 996, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 641, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:720)
root@inception:/home/andy# cat /etc/nagios/check_nrpe_ng.cfg
ssl_cert_file=/etc/ssl/certs/ssl-cert-snakeoil.pem
ssl_key_file=/etc/ssl/private/ssl-cert-snakeoil.key

ssl_verify_server=False
```

server end:

```
root@test:/home/andy# openssl x509 -text -in /etc/ssl/certs/ssl-cert-snakeoil.pem | grep CN
        Issuer: CN = test
        Subject: CN = test
```

**Comment by andydavidson** [comment]
I did look at this, creating a CA.  When there is a chain of trust, the CERTIFICATE_VERIFY_FAILED error goes away.  However, it looks like ssl_verify_server is still trying to verify the hostname and CN matches which I expect would not happen with ssl_verify_server = False:

```
root@inception:/home/andy# /usr/lib/nagios/plugins/check_nrpe_ng -H 192.168.56.103 -c check_load -C /etc/nagios/check_nrpe_ng.cfg --debug
[   ('args', None),
    ('command', 'check_load'),
    ('config_file', '/etc/nagios/check_nrpe_ng.cfg'),
    ('debug', True),
    ('host', '192.168.56.103'),
    ('port', 59546),
    ('ssl_ca_file', '/etc/ssl/certs/nrpe-ca.cert.pem'),
    ('ssl_cert_file', '/etc/ssl/certs/nrpe.pem'),
    ('ssl_key_file', '/etc/ssl/private/nrpe.key.pem'),
    ('ssl_verify_server', False),
    ('timeout', 10),
    ('timeout_unknown', False)]
Traceback (most recent call last):
  File "/usr/lib/nagios/plugins/check_nrpe_ng", line 11, in <module>
    load_entry_point('nrpe-ng==0.1.2', 'console_scripts', 'check_nrpe_ng')()
  File "/usr/lib/python3/dist-packages/nrpe_ng/client.py", line 197, in main
    return Client().run()
  File "/usr/lib/python3/dist-packages/nrpe_ng/client.py", line 177, in run
    conn.request(**req)
  File "/usr/lib/python3.5/http/client.py", line 1107, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.5/http/client.py", line 1152, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.5/http/client.py", line 1103, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/usr/lib/python3.5/http/client.py", line 877, in send
    self.connect()
  File "/usr/lib/python3.5/http/client.py", line 1261, in connect
    server_hostname=server_hostname)
  File "/usr/lib/python3.5/ssl.py", line 385, in wrap_socket
    _context=self)
  File "/usr/lib/python3.5/ssl.py", line 760, in __init__
    self.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 996, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 646, in do_handshake
    match_hostname(self.getpeercert(), self.server_hostname)
  File "/usr/lib/python3.5/ssl.py", line 309, in match_hostname
    % (hostname, dnsnames[0]))
ssl.CertificateError: hostname '192.168.56.103' doesn't match 'nrpe'

```

**Comment by bootc** [comment]
Thanks for reporting. Because you're running Debian 9 (`stretch`) you'll be running version 0.1.2 which doesn't actually do anything with the value of the `ssl_verify_server` setting in the client, and therefore validation is effectively forced on. This is documented in the `TODO` file in that version.

This should all be fixed in version 0.2.0 but that isn't available in `stretch`. Because its implementation in 0.2.0 depends on a significant rewrite of the client, it's unfortunately not possible to backport that code onto the 0.1 branch for `stretch`.

I could, however, produce a backport of 0.2.0 for `stretch-backports`. Please let me know if that would be of interest.

**Comment by igorlidin** [comment]
It would be great to have 0.2.0 in debian stretch backports repo. I'm interested in this.
